sdk/appconfiguration/azure-spring-cloud-appconfiguration-config/src/main/java/com/azure/spring/cloud/config/implementation/AppConfigurationRefreshUtil.java

```diff
@@ -0,0 +1,333 @@
+     *
+     * @return If a refresh event is called.
+     */
+    static RefreshEventData refreshStoresCheck(AppConfigurationReplicaClientFactory clientFactory,
+        Duration refreshInterval, Long defaultMinBackoff) {
+        RefreshEventData eventData = new RefreshEventData();
+        BaseAppConfigurationPolicy.setWatchRequests(true);
+
+        try {
+            if (refreshInterval != null && StateHolder.getNextForcedRefresh() != null
+                && Instant.now().isAfter(StateHolder.getNextForcedRefresh())) {
+                String eventDataInfo = "Minimum refresh period reached. Refreshing configurations.";
+
+                LOGGER.info(eventDataInfo);
+
+                eventData.setMessage(eventDataInfo);
+            }
+            for (Entry<String, ConnectionManager> entry : clientFactory.getConnections().entrySet()) {
+
+                String originEndpoint = entry.getKey();
+                ConnectionManager connection = entry.getValue();
+                // For safety reset current used replica.
+                clientFactory.setCurrentConfigStoreClient(originEndpoint, originEndpoint);
+                AppConfigurationStoreMonitoring monitor = connection.getMonitoring();
+
+                List<AppConfigurationReplicaClient> clients = clientFactory.getAvailableClients(originEndpoint);
+
+                if (monitor.isEnabled() && StateHolder.getLoadState(originEndpoint)) {
+                    for (AppConfigurationReplicaClient client : clients) {
+                        try {
+                            refreshWithTime(client, StateHolder.getState(originEndpoint), monitor.getRefreshInterval(),
+                                eventData);
+                            if (eventData.getDoRefresh()) {
+                                clientFactory.setCurrentConfigStoreClient(originEndpoint, client.getEndpoint());
+                                return eventData;
+                            }
+                            // If check didn't throw an error other clients don't need to be checked.
+                            break;
+                        } catch (AppConfigurationStatusException e) {
+                            LOGGER.warn("Failed attempting to connect to " + client.getEndpoint()
+                                + " during refresh check.");
+
+                            clientFactory.backoffClient(originEndpoint, client.getEndpoint());
+                            continue;
+                        }
+
+                    }
+                } else {
+                    LOGGER.debug("Skipping configuration refresh check for " + originEndpoint);
+                }
+
+                FeatureFlagStore featureStore = connection.getFeatureFlagStore();
+
+                if (featureStore.getEnabled() && StateHolder.getLoadStateFeatureFlag(originEndpoint)) {
+                    for (AppConfigurationReplicaClient client : clients) {
```

Comment:
>for (AppConfigurationReplicaClient client : clients) [](http://example.com/codeflow?start=20&length=52)

It looks like we iterate over client list to load feature flags after loading everything else. Is it possible that we are loading key-values from one endpoint and feature flags from another endpoint since they are in different for loops?

Reply:
This isn't loading anything, just checking if watch keys have changed. In the case of feature flags, we check for added/deleted ones and modified ones.

If any replica returns a non error message it will break out of the loop.

Reply:
So if there is a failoverable error while monitoring watch keys, we just check again with the next replica client. Full reload will happen later with whichever client is available at that time. 
Is my understanding correct?

Reply:
The Logic is this

1. Check watch key
2. If change, record current client & trigger refresh. If no change, nothing. If error, check with next client by going to step 1.
3. Refresh attempt with recorded current client. If success you have new config. If fail, we check the next client to see if it's watch key is updated. If it is we continue the refresh with is. If not, move to next client. If all clients fail or aren't updated kill refresh attempt.